### ISSUE TITLE
High runtime when executing workflow on metagenomic data, especially in the centrifuge step

### ISSUE THREAD

**Issue by JulianMohr** [epi2me-labs/wf-metagenomics#3]
Hello community,

I recently performed a metagenomic analysis on sequencing data generated from a zymomock community and tested this nextflow workflow.

Unfortunately, I had the issue that the complete execution of the workflow required exceptional high runtime compared to the corresponding EPI2ME Labs Notebook ['Metagenomic classification tutorial'](https://github.com/epi2me-labs/tutorials). In particular, it seems that the `centrifuge` step is the reason for this behaviour.

This is the command, which I have used to run the workflow:

```
nextflow run epi2me-labs/wf-metagenomics \
        -w output/workspace \
        -profile standard \
        --fastq ../fastq_pass \
        --db_path db_store/zymo \
        --db_prefix zymo \
        --out_dir output \
        --wfversion latest \
        --threads 16
```

I used the `centrifuge` database files from the `test_data/` directory of this repo, which can be found [here](https://github.com/epi2me-labs/wf-metagenomics/blob/master/test_data/db_store/zymo.tar.gz).

About my environment:

```
  Nextflow Version: 21.04.3 build 5560
  Created: 21-07-2021 15:09 UTC (17:09 CEST)
  System: Linux 5.4.0-84-generic
  Runtime: Groovy 3.0.7 on OpenJDK 64-Bit Server VM 11.0.11+9-Ubuntu-0ubuntu2.18.04
  Encoding: UTF-8 (UTF-8)
  CPUs: 32 - Mem: 125.8 GB (14.6 GB) - Swap: 2 GB (1.4 GB)

  Launching `epi2me-labs/wf-metagenomics` [stoic_edison] - revision: ef8303bb01 [master]
```

The workflow terminated successfully after 3 h and 29 min:

```
executor >  local (11)
[82/24dd36] process > pipeline:fastcat (1)        [100%] 1 of 1 ✔  
[4d/c2894e] process > pipeline:centrifuge (1)     [100%] 1 of 1 ✔ 
[09/bf6e7f] process > pipeline:generateMaster (1) [100%] 1 of 1 ✔ 
[86/1ddea2] process > pipeline:splitByMaster (1)  [100%] 1 of 1 ✔
[a1/b957be] process > output (7)                  [100%] 7 of 7 ✔

Completed at: 08-Oct-2021 18:34:51
Duration    : 3h 29m
CPU hours   : 3.3
Succeeded   : 11
```

According to the 'Process execution timeline' that nextlfow automatically generates in the `output/execution/` directory for every run, the most time demanding step was the execution of `centrifuge` (with almost 3 h):

![process_execution_timeline](https://user-images.githubusercontent.com/81371922/137115271-9084fab1-8f1b-4cf9-bdd8-607e14242ec6.png)

In contrast, the EPI2ME Labs Notebook only required about 6 min for the `centrifuge` step on the same dataset with equal `centrifuge` database files when running with 16 threads. Therefore, it seems like this workflow is facing some critical performance issues.

I have also checked the number of threads generated for the execution of `centrifuge` during the run and found 16 (like specified in the command using the `--threads` flag). However, the average CPU usage of `centrifuge` for each core was very low (< 10%). When running the EPI2ME Labs Notebook, the CPU usage was at nearly 100 % for the `centrifuge` step.

Until now, I was not able to identify the problem. Do you have any suggestions?

**Comment by sagrudd** [comment]
@JulianMohr thanks for the feedback.

The centrifuge based workflow has been replaced with a revised, updated and improved workflow that now uses the kraken2 software for the taxonomic assignment of sequences. 

The nextflow software is rather strict in its allocation of resources - the default parameters that are specified in the workflows are appropriate for running analyses on a GridION (whilst sequencing) and the differences observed without some additional tuning are not surprising. 

Please have a look at the updated workflow and I would be delighted to help you through the process of preparing a minimal configuration file that will help you exploit the computational resources that you have available.

**Comment by JulianMohr** [comment]
Hello @sagrudd,

thank you for your assistance! Actually, I used the EPI2ME Labs notebook instead of the workflow for further analyses. However, it is very good to know that the workflow has been updated since I opened this issue.

I will evaluate the updated workflow when analyzing metagenomic data in the future and close this issue in the meanwhile. If I still face any performance issues, I will keep you posted.
